subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveIncrementalChangesTaskExecuter.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.execution;
+
+import org.gradle.api.Describable;
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.tasks.TaskExecuter;
+import org.gradle.api.internal.tasks.TaskExecuterResult;
+import org.gradle.api.internal.tasks.TaskExecutionContext;
+import org.gradle.api.internal.tasks.TaskStateInternal;
+import org.gradle.internal.change.Change;
+import org.gradle.internal.change.ChangeVisitor;
+import org.gradle.internal.change.DescriptiveChange;
+import org.gradle.internal.execution.history.AfterPreviousExecutionState;
+import org.gradle.internal.execution.history.BeforeExecutionState;
+import org.gradle.internal.execution.history.changes.DefaultExecutionStateChanges;
+import org.gradle.internal.execution.history.changes.ExecutionStateChanges;
+
+import javax.annotation.Nullable;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Resolves the incremental changes to pass to the task actions.
+ *
+ * @see org.gradle.api.tasks.incremental.IncrementalTaskInputs
+ */
+public class ResolveIncrementalChangesTaskExecuter implements TaskExecuter {
+    private final TaskExecuter delegate;
+
+    public ResolveIncrementalChangesTaskExecuter(TaskExecuter delegate) {
+        this.delegate = delegate;
+    }
+
+    @Nullable
+    @Override
+    public TaskExecuterResult execute(final TaskInternal task, TaskStateInternal state, final TaskExecutionContext context) {
+        ExecutionStateChanges changes = context.getTaskExecutionMode().getRebuildReason()
+            .map(new Function<String, ExecutionStateChanges>() {
+                @Override
+                public ExecutionStateChanges apply(String rebuildReason) {
+                    return new RebuildExecutionStateChanges(rebuildReason);
+                }
+            }).orElseGet(new Supplier<ExecutionStateChanges>() {
+                @Nullable
+                @Override
+                public ExecutionStateChanges get() {
+                    final AfterPreviousExecutionState afterPreviousExecution = context.getAfterPreviousExecution();
```

Comment:
Shouldn't `AfterPreviousExecution` also be an Optional? So we could use `filter` and `flatMap`?

Reply:
It probably should. But: baby steps! I think all these changes will be both more obvious and easier to make when have moved things over to `:execution`.